src/rust/kafka-executor/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+        producer,
+        deserializer,
+        serializer,
+    ),
+    fields(
+        src_topic_name = %consumer.topic_name,
+        dst_topic_name = %producer.topic_name,
+    )
+)]
+pub async fn service_loop<
+    InputEventT: Send + Sync + 'static,
+    OutputEventT: Send + Sync + 'static,
+    EventHandlerErrorT: std::error::Error + Send + Unpin + 'static,
+    EventHandlerT: EventHandler<
+            InputEvent = InputEventT,
+            OutputEvent = OutputEventT,
+            Error = EventHandlerErrorT,
+        > + Send
+        + Unpin
+        + 'static,
+    EventDeserializerT: event_decoder::EventDeserializer<InputEvent = InputEventT> + Send + Unpin + 'static,
+    EventSerializerT: event_encoder::EventSerializer<OutputEvent = OutputEventT> + Send + Unpin + 'static,
+>(
+    event_handler: EventHandlerT,
+    consumer: EventConsumer<DefaultConsumerContext, DefaultRuntime>,
+    producer: EventProducer,
+    deserializer: EventDeserializerT,
+    serializer: EventSerializerT,
+) {
+    pin_mut!(event_handler);
+    pin_mut!(deserializer);
+    pin_mut!(serializer);
+    let producer = Pin::new(&producer);
+
+    while let Some(message) = consumer.consumer.stream().next().await {
+        let message = match message {
+            Ok(message) => message,
+            Err(e) => {
+                tracing::error!(
+                    message="Failed to read kafka message",
+                    error=%e,
+                );
+                continue;
+            }
+        };
+        process_message(
+            message,
+            event_handler.as_mut(),
+            &consumer,
+            producer.as_ref(),
+            deserializer.as_mut(),
+            serializer.as_mut(),
+        )
+        .await;
+    }
+}
+
+#[tracing::instrument(
+    skip(
+        message,
+        event_handler,
+        consumer,
+        producer,
+        deserializer,
+        serializer,
+    ),
+    fields(
+        offset = message.offset(),
+        key = ?message.key()
+    )
+)]
+async fn process_message<
+    InputEventT: Send + 'static,
+    OutputEventT: Send + 'static,
+    EventHandlerErrorT: std::error::Error + Send + 'static,
+    EventHandlerT: EventHandler<
+            InputEvent = InputEventT,
+            OutputEvent = OutputEventT,
+            Error = EventHandlerErrorT,
+        > + Send
+        + Unpin
+        + 'static,
+    EventDeserializerT: event_decoder::EventDeserializer<InputEvent = InputEventT> + Unpin + Send + 'static,
+    EventSerializerT: event_encoder::EventSerializer<OutputEvent = OutputEventT> + Unpin + Send + 'static,
+>(
+    message: BorrowedMessage<'_>,
+    mut event_handler: Pin<&mut EventHandlerT>,
+    consumer: &EventConsumer<DefaultConsumerContext, DefaultRuntime>,
+    producer: Pin<&EventProducer>,
+    mut deserializer: Pin<&mut EventDeserializerT>,
+    mut serializer: Pin<&mut EventSerializerT>,
+) {
+    let event = match deserializer.deserialize(message.payload().unwrap()) {
```

Comment:
This assumption feels a little less naunced than it needs to be. We're going to have an Envelope event which _should_ be deserialized but the _inner_ event _shouldn't_ be deserialized (unless the caller requests it).

EDIT: maybe we can express this in the `Deserializer`?

Reply:
I would imagine that that's expressed in the Deserializer, yeah. If you want to just pass it along, your Deserializer would only deserialize the Envelope.

Reply:
We should probably unconditionally deserialize the Envelope in the process_message function, and then allow the Deserializer to optionally deserialize the internal bytes (it can just hand bytes back if it isn't interested).